plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar310/TreeBuilder310.java

```diff
@@ -632,76 +634,250 @@ public final SimpleNode onCloseNode(SimpleNode n, int arity) throws Exception {
                     } catch (Exception e) {
                         Log.log("Expected expr. Found: " + popNode);
                     }
+                    return new MatchMapping(new exprType[] { key }, pattern, null);
                 } else {
-                    Log.log("Expected arity to be 1 here.");
+                    Log.log("Expected arity to be 1 or 2 here.");
                 }
-                return expr;
+                return null;
 
-            case JJTAS_PATTERN:
-                exprType pattern = null;
-                String name = null;
-                if (arity == 2) {
+            case JJTMAPPING_PATTERN:
+                if (arity > 0) {
+                    exprType[] keys = new exprType[arity];
+                    patternType[] patterns = new patternType[arity];
+                    MatchMapping matchMapping = null;
+                    for (int i = 0; i < arity; i++) {
+                        SimpleNode popNode = stack.popNode();
+                        try {
+                            matchMapping = (MatchMapping) popNode;
+                        } catch (Exception e) {
+                            Log.log("Expected expr. Found: " + popNode);
+                        }
+                        keys[i] = matchMapping.keys[0];
+                        patterns[i] = matchMapping.patterns[0];
+                    }
+                    return new MatchMapping(keys, patterns, null);
+                }
+                return new MatchMapping(new exprType[0], new patternType[0], null);
+
+            case JJTSEQUENCE_PATTERN:
+                if (arity > 0) {
+                    patternType[] patterns = new patternType[arity];
+                    for (int i = 0; i < arity; i++) {
+                        SimpleNode popNode = stack.popNode();
+                        try {
+                            patterns[i] = (patternType) popNode;
+                        } catch (Exception e) {
+                            Log.log("Expected pattern. Found: " + popNode);
+                        }
+                    }
+                    return new MatchSequence(patterns);
+                }
+                return new MatchSequence(new patternType[0]);
+
+            case JJTATTR:
+                if (arity >= 2) {
+                    SimpleNode popNode = stack.popNode();
+                    if (popNode instanceof patternType) {
+                        if (arity == 2) {
+                            patternType pattern = null;
+                            try {
+                                pattern = (patternType) popNode;
+                            } catch (Exception e) {
+                                Log.log("Expected pattern. Found: " + popNode);
+                            }
+                            popNode = stack.popNode();
+                            String rep = null;
+                            try {
+                                rep = NodeUtils.getFullRepresentationString(popNode);
```

Comment:
This isn't ideal, we should try to leave the same node that was originally constructed (as later on, when pretty-printing it can make a difference to reconstruct the code from the nodes).